### PR TITLE
Don't attempt copy/remove fallback for rename failures

### PR DESCRIPTION
**Pull request by nehaljwani**
Drop custom copy + remove fallback for rename failures

`os.rename()` returns `EXDEV` for directories and `EINVAL` for files in a docker
container with `overlayfs` or `aufs` storage backend.

xref: https://github.com/moby/moby/issues/25409#issuecomment-238537855

The fallback for this was to use copy + remove, but it assumed that the rename
failed for a file. So, whenever a copy attempt was made on a directory, it
cascaded into an uncaught `EISDIR`.

This patch delegates the copy+remove logic to `shutil.move()` which works for
directories and files and saves us from reinventing the wheel.

### CODE REVIEW

File: conda/gateways/disk/update.py [hunk 1]
```diff
@@ -62,6 +62,11 @@ def rename(source_path, destination_path, force=False):
             os_rename(source_path, destination_path)
         except EnvironmentError as e:
             if e.errno in (EINVAL, EXDEV):
+                # https://github.com/conda/conda/issues/6811
+                # Don't attempt fallback for directories, let the caller catch EXDEV, see:
+                # https://github.com/conda/conda/blob/4.4.9/conda/core/path_actions.py#L1243-L1254
+                if e.errno == EXDEV and isdir(source_path):
+                    raise
```
Comment (nehaljwani) [inline conda/gateways/disk/update.py]: To be honest, I think we should just get rid of this fallback and add EINVAL to https://github.com/conda/conda/blob/4.4.9/conda/core/path_actions.py#L1246

But I am not sure of rename is being called only from that place or not.

Comment (nehaljwani) [reply conda/gateways/disk/update.py]: cc @quasiben

Comment (kalefranz) [reply conda/gateways/disk/update.py]: `rename()` is called 7 times through `backoff_rename()` in `conda/core/path_actions.py`. It looks like that's currently the full use of it.

Comment (kalefranz) [reply conda/gateways/disk/update.py]: When using `backoff_rename()` in a `reverse()` method, I don't think we have to worry about any special handling for docker filesystem issues.

Use of `backoff_rename()` in `execute()` methods at:

* https://github.com/conda/conda/blob/4.4.9/conda/core/path_actions.py#L974
* https://github.com/conda/conda/blob/4.4.9/conda/core/path_actions.py#L1151
* https://github.com/conda/conda/blob/4.4.9/conda/core/path_actions.py#L1244

Comment (kalefranz) [reply conda/gateways/disk/update.py]: > I think we should just get rid of this fallback completely and add EINVAL to 

I'm ok with raising the error handling up to the caller of `rename()`/`backoff_rename()` for now.  @nehaljwani Why don't you go ahead and see what that looks like.  Just make sure to add notes in top-line comments for `rename()` and `backoff_rename()` about the errors future users of those functions might want to think about handling.

**Comment by kalefranz** [comment]
Wasn't there an issue this was meant to address? I thought so, but I'm not finding it...

**Comment by nehaljwani** [comment]
@kalefranz https://github.com/conda/conda/issues/6811

**Comment by kalefranz** [comment]
@nehaljwani Do you think we should backport this to 4.3.x?

**Comment by nehaljwani** [comment]
We could, yes.

**Pull request merged**